diff --git a/debian/rules b/debian/rules
index b23b335..d37413c 100755
--- a/debian/rules
+++ b/debian/rules
@@ -31,10 +31,6 @@ endif
 %:
 	dh $@
 
-# The net-snmp Makefiles cannot handle parallel builds
-override_dh_auto_build:
-	dh_auto_build --no-parallel
-
 override_dh_auto_configure:
 	dh_auto_configure -- --prefix=/usr --sysconfdir=/etc --mandir=/usr/share/man \
 	  --with-persistent-directory=/var/lib/snmp \
-- 
2.34.1

