diff --git a/makefile b/makefile
index f23f0ac..a16b2ce 100644
--- a/makefile
+++ b/makefile
@@ -93,10 +93,10 @@ V ?= 1
 
 ifneq ($(findstring 3.10,$(KVERSION))$(findstring 3.13,$(KVERSION))$(findstring 3.14,$(KVERSION))$(findstring 3.16,$(KVERSION)),)
 MLNX_LINUX_AUTOCONF_FILE = include/generated/autoconf.h
-MLNX_LINUX_EXTRA_INCLUDE_FILES = -include include/linux/kconfig.h
+MLNX_LINUX_EXTRA_INCLUDE_FILES = -include $(KSRC_EXT)/include/linux/kconfig.h
 MLNX_LINUX_EXTRA_INCLUDE_FOLDERS = \
-		-Iarch/$$(SRCARCH)/include/uapi \
-		-Iinclude/uapi \
+		-I$(KSRC_EXT)/arch/$$(SRCARCH)/include/uapi \
+		-I$(KSRC_EXT)/include/uapi \
 		-Iarch/$$(SRCARCH)/include/generated/uapi \
 		-Iarch/$$(SRCARCH)/include/generated \
 		-Iinclude/generated/uapi
