From db633987abf1ea093cb1785b3cd37adfdb55e4cc Mon Sep 17 00:00:00 2001
From: Qi Luo <qiluo-msft@users.noreply.github.com>
Date: Mon, 15 Oct 2018 22:30:28 +0000
Subject: [PATCH] Disable SNMPv1

Signed-off-by: Qi Luo <qiluo-msft@users.noreply.github.com>
---
 debian/rules | 1 +
 1 file changed, 1 insertion(+)

diff --git a/debian/rules b/debian/rules
index 9eb8d2d..4c3b5b6 100755
--- a/debian/rules
+++ b/debian/rules
@@ -42,6 +42,7 @@ endif
 override_dh_auto_configure:
 	dh_auto_configure -- --prefix=/usr --sysconfdir=/etc --mandir=/usr/share/man \
 	  --with-persistent-directory=/var/lib/snmp \
+	  --disable-snmpv1 \
 	  --enable-ucd-snmp-compatibility \
 	  --enable-shared --with-cflags="$(CFLAGS) -DNETSNMP_USE_INLINE" \
 	  --with-ldflags="$(LDFLAGS)" \
-- 
2.18.0

