From e10409ea82c16c0e48316a0b4d71a0f649f92a54 Mon Sep 17 00:00:00 2001
From: liora <liora@nvidia.com>
Date: Mon, 17 Apr 2023 05:17:27 +0000
Subject: [PATCH 1/1] Fix version identifier

---
 include/version.h | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/include/version.h b/include/version.h
index 89d0597..cf7afbb 100644
--- a/include/version.h
+++ b/include/version.h
@@ -1 +1 @@
-static const char version[] = "5.9.0";
+static const char version[] = "5.10.0";
-- 
2.30.2

