Bug 1970 UNLINK_EXPR_SLIST() causes crash if list is empty

From: Arun Barboza <arun.barboza@broadcom.com>


---
 include/ntp_lists.h |    6 +++++-
 1 file changed, 5 insertions(+), 1 deletion(-)

diff --git a/include/ntp_lists.h b/include/ntp_lists.h
index d741974..f90bf23 100644
--- a/include/ntp_lists.h
+++ b/include/ntp_lists.h
@@ -184,7 +184,11 @@ do {								\
 do {								\
 	entrytype **ppentry;					\
 								\
-	ppentry = &(listhead);					\
+	if (!listhead) {					\
+		(punlinked) = NULL;				\
+		break;						\
+	}							\
+	else ppentry = &(listhead);				\
 								\
 	while (!(expr))						\
 		if (*ppentry != NULL &&				\
