--- a/patch/series
+++ b/patch/series
@@ -51,6 +51,10 @@
 0105-mlxsw-reg-Extend-MTBR-register-with-new-slot-number-.patch
 0106-mlxsw-reg-Extend-MCIA-register-with-new-slot-number-.patch
 0107-mlxsw-reg-Extend-MCION-register-with-new-slot-number.patch
+0168-TMP-mlxsw-minimal-Ignore-error-reading-SPAD-register.patch
+0169-TMP-mlxsw-i2c-Prevent-transaction-execution-for-spec.patch
+0172-DS-platform-mlx-platform-Add-SPI-path-for-rack-switc.patch
+0174-DS-mlxsw-core_linecards-Skip-devlink-and-provisionin.patch
 ###-> mellanox_hw_mgmt-end
 
 # Cisco patches for 5.10 kernel
