diff --git a/teamd/teamd_runner_lacp.c b/teamd/teamd_runner_lacp.c
index d6e5840..2144334 100644
--- a/teamd/teamd_runner_lacp.c
+++ b/teamd/teamd_runner_lacp.c
@@ -1096,6 +1096,7 @@ static int lacpdu_recv(struct lacp_port *lacp_port)
 	struct lacpdu lacpdu;
 	struct sockaddr_ll ll_from;
 	int err;
+	bool admin_state;
 
 	err = teamd_recvfrom(lacp_port->sock, &lacpdu, sizeof(lacpdu), 0,
 			     (struct sockaddr *) &ll_from, sizeof(ll_from));
@@ -1105,6 +1106,10 @@ static int lacpdu_recv(struct lacp_port *lacp_port)
 	if (!teamd_port_present(lacp_port->ctx, lacp_port->tdport))
 		return 0;
 
+	admin_state = team_get_ifinfo_admin_state(lacp_port->ctx->ifinfo);
+	if (!admin_state)
+		return 0;
+
 	if (!lacpdu_check(&lacpdu)) {
 		teamd_log_warn("malformed LACP PDU came.");
 		return 0;
