From 63b5b14ad289f18928beac65754e7bb13183b5dc Mon Sep 17 00:00:00 2001
From: Pavel Shirshov <pavelsh@microsoft.com>
Date: Mon, 4 Nov 2019 18:12:54 -0800
Subject: [PATCH 1/1] Reduce severity of 'Vty connected from' message

---
 lib/vty.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/lib/vty.c b/lib/vty.c
index deb9391bd..743ff1c17 100644
--- a/lib/vty.c
+++ b/lib/vty.c
@@ -1853,7 +1853,7 @@ static int vty_accept(struct thread *thread)
 		zlog_info("can't set sockopt to vty_sock : %s",
 			  safe_strerror(errno));
 
-	zlog_info("Vty connection from %s",
+	zlog_debug("Vty connection from %s",
 		  sockunion2str(&su, buf, SU_ADDRSTRLEN));
 
 	vty_create(vty_sock, &su);
-- 
2.17.1.windows.2

