diff --git a/common/discover.c b/common/discover.c
index 3cd64a7..c85d18c 100644
--- a/common/discover.c
+++ b/common/discover.c
@@ -547,7 +547,7 @@ next_iface4(struct iface_info *info, int *err, struct iface_conf_list *ifaces) {
 				log_error("Interface name '%s' too long", name);
 				return 0;
 			}
-			strncpy(info->name, name, sizeof(info->name) - 1);
+			strncpy(info->name, name, sizeof(info->name));
 
 #ifdef ALIAS_NAMED_PERMUTED
 			/* interface aliases look like "eth0:1" or "wlan1:3" */
