From 7a8d588fa0425550cfd42c3300c8c0a643137190 Mon Sep 17 00:00:00 2001
From: Myron Sosyak <myronx.sosyak@intel.com>
Date: Fri, 13 Aug 2021 11:26:13 +0000
Subject: [PATCH] Remove-unneeded-packages

---
 debian/control | 28 ----------------------------
 debian/rules   |  7 +++----
 2 files changed, 3 insertions(+), 32 deletions(-)

diff --git a/debian/control b/debian/control
index 5126e47..3f3028c 100644
--- a/debian/control
+++ b/debian/control
@@ -13,8 +13,6 @@ Build-Depends: debhelper-compat (= 11), dh-python,
  python3-all-dev,
  python3-six,
  golang-go, golang-github-golang-mock-dev,
- pkg-php-tools (>= 1.14~), php-dev, phpunit, php-codesniffer,
- perl (>= 5.22), libbit-vector-perl, libclass-accessor-perl,
 # openjdk-11-jdk, javahelper, maven-debian-helper (>= 1.5), ant (>= 1.7), ant-optional,
 # libhttpclient-java, libslf4j-java, libservlet3.1-java (>= 8),
 # nodejs, npm,
@@ -175,32 +173,6 @@ Description: Python library for Thrift (debug symbols)
 # .
 # This package contains Javadoc HTML documentation for libthrift-java .
 #
-Package: php-thrift
-Architecture: any
-Section: php
-Depends: ${phpcomposer:Debian-require}, ${shlibs:Depends}, ${misc:Depends}
-Provides: ${php:Provides}
-Conflicts: php7.0-thrift, php7.2-thrift
-Replaces: php7.0-thrift, php7.2-thrift
-Description: PHP language support for Thrift
- Thrift is a software framework for the development of reliable and
- performant communication and data serialization. It combines a software
- stack with code generation to build services that operate seamlessly
- across a number of different development languages.
- .
- This package provides the PHP5 language support for Thrift.
-
-Package: libthrift-perl
-Section: perl
-Architecture: all
-Depends: ${perl:Depends}, ${misc:Depends}
-Description: Perl language support for Thrift
- Thrift is a software framework for the development of reliable and
- performant communication and data serialization. It combines a software
- stack with code generation to build services that operate seamlessly
- across a number of different development languages.
- .
- This package provides the Perl language support for Thrift.

 Package: golang-thrift-dev
 Architecture: amd64 arm64 armel armhf i386 ppc64el
diff --git a/debian/rules b/debian/rules
index ba63e6f..8217796 100755
--- a/debian/rules
+++ b/debian/rules
@@ -47,7 +47,6 @@ override_dh_auto_clean:
 		$(CURDIR)/lib/go/test/gopath/src/golang.org/x/net \
 		$(CURDIR)/lib/go/test/gopath/src/github.com/golang/mock
 	[ ! -f Makefile ] || $(MAKE) clean
-	cd $(CURDIR)/lib/php/src/ext/thrift_protocol/ && phpize --clean
 	find $(CURDIR) \( -name Makefile -o -name Makefile.in \) -exec rm {} \;
 	find $(CURDIR) -name .dirstamp -exec rm {} \;
 	$(RM) aclocal.m4 config.h config.log config.status configure \
@@ -65,10 +64,10 @@ override_dh_auto_clean:

 override_dh_auto_configure:
 	dh_auto_configure -- \
-		--with-cpp --with-c_glib --with-php --with-php_extension \
-		--with-perl --without-python --with-py3 --without-java \
+		--with-cpp --with-c_glib --without-php --without-php_extension \
+		--without-perl --without-python --with-py3 --without-java \
 		--without-nodejs --without-erlang --without-lua \
-		--with-go --with-ruby \
+		--without-go --without-ruby \
 		--with-qt4=no --with-qt5=yes \
 		--without-rust \
 		--disable-tutorial
--
2.17.1

