From 15e692fb82a61203624829cdd872315211920077 Mon Sep 17 00:00:00 2001
From: Guohan Lu <gulv@microsoft.com>
Date: Tue, 6 Mar 2018 09:36:51 +0000
Subject: [PATCH] return error when port does not exist

---
 src/client/conf-lldp.c | 8 ++++++++
 1 file changed, 8 insertions(+)

diff --git a/src/client/conf-lldp.c b/src/client/conf-lldp.c
index c16219b..b51e4eb 100644
--- a/src/client/conf-lldp.c
+++ b/src/client/conf-lldp.c
@@ -148,6 +148,8 @@ cmd_portid_type_local(struct lldpctl_conn_t *conn, struct writer *w,
 	const char *name;
 	const char *id = cmdenv_get(env, "port-id");
 	const char *descr = cmdenv_get(env, "port-descr");
+	const char *portname = cmdenv_get(env, "ports");
+	int find_port = 0;
 
 	log_debug("lldpctl", "lldp PortID TLV Subtype Local port-id '%s' port-descr '%s'", id, descr);
 
@@ -165,6 +167,12 @@ cmd_portid_type_local(struct lldpctl_conn_t *conn, struct writer *w,
 			log_warnx("lldpctl", "unable to set LLDP Port Description for %s."
 			    " %s", name, lldpctl_last_strerror(conn));
 		}
+		find_port = 1;
+	}
+
+	if (!find_port) {
+		log_warnx("lldpctl", "cannot find port %s", portname);
+		return 0;
 	}
 
 	return 1;
-- 
2.7.4


