From 2ee8585e8b716719a11235ab5c291b2f6ac9ce1a Mon Sep 17 00:00:00 2001
From: Pavel Shirshov <pavelsh@microsoft.com>
Date: Wed, 17 Oct 2018 21:05:58 +0000
Subject: [PATCH] Let linux kernel to find appropriate nl_pid automatically

---
 src/daemon/netlink.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/src/daemon/netlink.c b/src/daemon/netlink.c
index 1a64a23..f4227b8 100644
--- a/src/daemon/netlink.c
+++ b/src/daemon/netlink.c
@@ -93,7 +93,7 @@ netlink_connect(struct lldpd *cfg, int protocol, unsigned groups)
 	int s;
 	struct sockaddr_nl local = {
 		.nl_family = AF_NETLINK,
-		.nl_pid = getpid(),
+		.nl_pid = 0,
 		.nl_groups = groups
 	};
 
-- 
2.7.4

