From 833653dffb9be40110142af2a7cb4076a0dd24f5 Mon Sep 17 00:00:00 2001
From: Pavel Shirshov <pavelsh@microsoft.com>
Date: Thu, 12 Dec 2019 12:48:47 -0800
Subject: [PATCH 1/1] lib: fix memory leak

---
 src/lib/connection.c | 1 +
 1 file changed, 1 insertion(+)

diff --git a/src/lib/connection.c b/src/lib/connection.c
index 88bbc99..aa88dad 100644
--- a/src/lib/connection.c
+++ b/src/lib/connection.c
@@ -114,6 +114,7 @@ lldpctl_new_name(const char *ctlname, lldpctl_send_callback send, lldpctl_recv_c
 	}
 	if (!send && !recv) {
 		if ((data = malloc(sizeof(struct lldpctl_conn_sync_t))) == NULL) {
+			free(conn->ctlname);
 			free(conn);
 			return NULL;
 		}
-- 
2.17.1.windows.2

