From fdb789c348fdcde6d5ef8b837d7f33718bc0862b Mon Sep 17 00:00:00 2001
From: sudhanshukumar22 <sudhanshu.kumar@broadcom.com>
Date: Mon, 23 Nov 2020 20:47:28 -0800
Subject: [PATCH] Ported fix for https://github.com/lldpd/lldpd/issues/408 from
 community

---
 src/lib/atom.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/src/lib/atom.c b/src/lib/atom.c
index f81d3bb..75c1275 100644
--- a/src/lib/atom.c
+++ b/src/lib/atom.c
@@ -327,7 +327,7 @@ _lldpctl_do_something(lldpctl_conn_t *conn,
 			conn->state_data[0] = 0;
 	}
 	if (conn->state == state_send &&
-	    (state_data == NULL || !strncmp(conn->state_data, state_data, sizeof(conn->state_data)))) {
+	    (state_data == NULL || !strncmp(conn->state_data, state_data, sizeof(conn->state_data) - 1))) {
 		/* We need to send the currently built message */
 		rc = lldpctl_send(conn);
 		if (rc < 0)
@@ -335,7 +335,7 @@ _lldpctl_do_something(lldpctl_conn_t *conn,
 		conn->state = state_recv;
 	}
 	if (conn->state == state_recv &&
-	    (state_data == NULL || !strncmp(conn->state_data, state_data, sizeof(conn->state_data)))) {
+	    (state_data == NULL || !strncmp(conn->state_data, state_data, sizeof(conn->state_data) - 1))) {
 		/* We need to receive the answer */
 		while ((rc = ctl_msg_recv_unserialized(&conn->input_buffer,
 			    &conn->input_buffer_len,
-- 
2.12.2

