From 10ed7b7e0fd06c312f497e4647c4140b1cc8eef8 Mon Sep 17 00:00:00 2001
From: Guohan Lu <gulv@microsoft.com>
Date: Mon, 8 May 2017 05:53:38 +0000
Subject: [PATCH] add PHY support for Broadcom 54616

---
 src/e1000_82575.c   | 5 +++++
 src/e1000_defines.h | 1 +
 src/e1000_hw.h      | 1 +
 3 files changed, 7 insertions(+)

diff --git a/src/e1000_82575.c b/src/e1000_82575.c
index b4b973e..afaa7e5 100644
--- a/src/e1000_82575.c
+++ b/src/e1000_82575.c
@@ -302,6 +302,9 @@ static s32 e1000_init_phy_params_82575(struct e1000_hw *hw)
 		phy->ops.set_d3_lplu_state = e1000_set_d3_lplu_state_82580;
 		phy->ops.force_speed_duplex = e1000_phy_force_speed_duplex_m88;
 		break;
+	case BCM54616_E_PHY_ID:
+		phy->type = e1000_phy_bcm54616;
+		break;
 	default:
 		ret_val = -E1000_ERR_PHY;
 		goto out;
@@ -1602,6 +1605,8 @@ static s32 e1000_setup_copper_link_82575(struct e1000_hw *hw)
 	case e1000_phy_82580:
 		ret_val = e1000_copper_link_setup_82577(hw);
 		break;
+	case e1000_phy_bcm54616:
+		break;
 	default:
 		ret_val = -E1000_ERR_PHY;
 		break;
diff --git a/src/e1000_defines.h b/src/e1000_defines.h
index 6de3988..773c464 100644
--- a/src/e1000_defines.h
+++ b/src/e1000_defines.h
@@ -1185,6 +1185,7 @@
 #define I210_I_PHY_ID		0x01410C00
 #define IGP04E1000_E_PHY_ID	0x02A80391
 #define M88_VENDOR		0x0141
+#define BCM54616_E_PHY_ID	0x03625D10
 
 /* M88E1000 Specific Registers */
 #define M88E1000_PHY_SPEC_CTRL		0x10  /* PHY Specific Control Reg */
diff --git a/src/e1000_hw.h b/src/e1000_hw.h
index 3bcecf1..cca5d77 100644
--- a/src/e1000_hw.h
+++ b/src/e1000_hw.h
@@ -133,6 +133,7 @@ enum e1000_phy_type {
 	e1000_phy_82580,
 	e1000_phy_vf,
 	e1000_phy_i210,
+	e1000_phy_bcm54616,
 };
 
 enum e1000_bus_type {
-- 
2.7.4

