diff --git a/man/teamd.conf.5 b/man/teamd.conf.5
index 350ffc9..dc913cd 100644
--- a/man/teamd.conf.5
+++ b/man/teamd.conf.5
@@ -248,7 +248,7 @@ Default:
 .RE
 .TP
 .BR "runner.min_ports " (int)
-Specifies the minimum number of ports that must be active before asserting carrier in the master interface, value can be 1 \(en 255.
+Specifies the minimum number of ports that must be active before asserting carrier in the master interface, value can be 1 \(en 1024.
 .RS 7
 .PP
 Default:
diff --git a/teamd/teamd_runner_lacp.c b/teamd/teamd_runner_lacp.c
index 9354ebb..a901398 100644
--- a/teamd/teamd_runner_lacp.c
+++ b/teamd/teamd_runner_lacp.c
@@ -151,6 +151,7 @@ struct lacp {
 #define		LACP_CFG_DFLT_FALLBACK false
 		int min_ports;
 #define		LACP_CFG_DFLT_MIN_PORTS 1
+#define		LACP_CFG_DFLT_MIN_PORTS_MAX 1024
 		enum lacp_agg_select_policy agg_select_policy;
 #define		LACP_CFG_DFLT_AGG_SELECT_POLICY LACP_AGG_SELECT_LACP_PRIO
 	} cfg;
@@ -493,7 +494,7 @@ static int lacp_load_config(struct teamd_context *ctx, struct lacp *lacp)
 	err = teamd_config_int_get(ctx, &tmp, "$.runner.min_ports");
 	if (err) {
 		lacp->cfg.min_ports = LACP_CFG_DFLT_MIN_PORTS;
-	} else if (tmp < 1 || tmp > UCHAR_MAX) {
+	} else if (tmp < 1 || tmp > LACP_CFG_DFLT_MIN_PORTS_MAX) {
 		teamd_log_err("\"min_ports\" value is out of its limits.");
 		return -EINVAL;
 	} else {
