From 99e50d2dbc9e855e457c03e19c4c3b894ffcaa7a Mon Sep 17 00:00:00 2001
From: Myron sosyak <myronx.sosyak@intel.com>
Date: Fri, 28 Jan 2022 15:02:16 +0000
Subject: [PATCH 2/2] Fix build rules

---
 debian/libthrift0.install |  1 -
 debian/rules              | 83 +++++++++++----------------------------
 2 files changed, 24 insertions(+), 60 deletions(-)

diff --git a/debian/libthrift0.install b/debian/libthrift0.install
index ec74efc5c..d4ca21e5e 100644
--- a/debian/libthrift0.install
+++ b/debian/libthrift0.install
@@ -1,5 +1,4 @@
 usr/lib/libthrift.so*
 usr/lib/libthrift-*.so*
 usr/lib/libthriftnb*.so*
-usr/lib/libthriftqt*.so*
 usr/lib/libthriftz*.so*
diff --git a/debian/rules b/debian/rules
index d0de50872..3a50319ee 100755
--- a/debian/rules
+++ b/debian/rules
@@ -16,7 +16,7 @@
 # This has to be exported to make some magic below work.
 export DH_OPTIONS

-PYVERS := $(shell pyversions -r)
+PYVERS := python2 python3

 export CPPFLAGS:=$(shell dpkg-buildflags --get CPPFLAGS)
 export CFLAGS:=$(shell dpkg-buildflags --get CFLAGS)
@@ -29,7 +29,13 @@ configure-stamp:

 	# Add here commands to configure the package.
 	if [ -f bootstrap.sh ]; then $(CURDIR)/bootstrap.sh; fi
-	$(CURDIR)/configure --prefix=/usr --enable-plugin=no
+	$(CURDIR)/configure --prefix=/usr --enable-plugin=no \
+		--with-cpp --with-c_glib --with-py3 --with-python \
+		--without-php_extension --without-perl \
+		--without-php --without-java \
+		--without-nodejs --without-erlang --without-lua \
+		--without-go --without-ruby --with-qt4=no \
+		--with-qt5=no --without-rust --disable-tutorial

 	touch configure-stamp

@@ -40,16 +46,7 @@ build: build-arch build-indep
 	# $(MAKE) -C test check

 build-arch: build-arch-stamp
-$(CURDIR)/compiler/cpp/thrift build-arch-stamp: configure-stamp
-
-	# Compile compiler
-	$(MAKE) -C $(CURDIR)/compiler/cpp
-
-	# Compile C++ library
-	$(MAKE) -C $(CURDIR)/lib/cpp
-
-	# Compile C (glib) library
-	$(MAKE) -C $(CURDIR)/lib/c_glib
+$(CURDIR)/compiler/cpp/thrift build-arch-stamp: configure-stamp build-libcpp-stamp build-libc-stamp build-compiler-stamp

 	# Python library
 	cd $(CURDIR)/lib/py && \
@@ -58,11 +55,22 @@ $(CURDIR)/compiler/cpp/thrift build-arch-stamp: configure-stamp
         $$py-dbg setup.py build; \
     done

-	# PHP
-	cd $(CURDIR)/lib/php/src/ext/thrift_protocol && \
-		phpize && \
-		./configure && $(MAKE)
+	touch $@

+build-compiler-stamp: configure-stamp
+	# Compile compiler
+	$(MAKE) -C $(CURDIR)/compiler/cpp
+	touch $@
+
+
+build-libcpp-stamp: configure-stamp build-compiler-stamp
+	# Compile C++ library
+	$(MAKE) -C $(CURDIR)/lib/cpp
+	touch $@
+
+build-libc-stamp: configure-stamp build-compiler-stamp
+	# Compile C (glib) library
+	$(MAKE) -C $(CURDIR)/lib/c_glib
 	touch $@

 build-indep: build-indep-stamp
@@ -71,19 +79,6 @@ build-indep-stamp: configure-stamp $(CURDIR)/compiler/cpp/thrift
 	# Add here commands to compile the indep part of the package.
 	#$(MAKE) doc

-	# Java
-	cd $(CURDIR)/lib/java && \
-		./gradlew --no-daemon -Prelease=true jar
-
-	# C#
-	$(MAKE) -C $(CURDIR)/lib/netstd
-
-	# Ruby
-	$(MAKE) -C $(CURDIR)/lib/rb
-
-	# Perl
-	$(MAKE) -C $(CURDIR)/lib/perl INSTALLDIRS=vendor
-
 	touch $@

 clean:
@@ -109,30 +104,6 @@ install-indep:
 	# debian/<package>-doc.
 	#INSTALLDOC#

-	# Java
-	mkdir -p $(CURDIR)/debian/libthrift-java/usr/share/java/ && \
-	cp $(CURDIR)/lib/java/build/libs/libthrift*.jar \
-		$(CURDIR)/debian/libthrift-java/usr/share/java/
-
-	# Ruby
-	mkdir -p $(CURDIR)/debian/ruby-thrift/usr/lib/ruby/1.9.1 && \
-	cp $(CURDIR)/lib/rb/lib/thrift.rb \
-		$(CURDIR)/debian/ruby-thrift/usr/lib/ruby/1.9.1
-	cp -r $(CURDIR)/lib/rb/lib/thrift \
-		$(CURDIR)/debian/ruby-thrift/usr/lib/ruby/1.9.1
-
-	# C#
-	mkdir -p $(CURDIR)/debian/libthrift-netstd/usr/lib/cli/thrift/ && \
-	cp $(CURDIR)/lib/netstd/Thrift/bin/Release/netstandard2.0/Thrift.dll \
-		$(CURDIR)/debian/libthrift-netstd/usr/lib/cli/thrift/Thrift.dll
-
-	# Perl
-	$(MAKE) -C $(CURDIR)/lib/perl install DESTDIR=$(CURDIR)/debian/libthrift-perl
-	mkdir -p $(CURDIR)/debian/libthrift-perl/usr/share
-	mv $(CURDIR)/debian/libthrift-perl/usr/local/lib/perl5 $(CURDIR)/debian/libthrift-perl/usr/share
-	rmdir $(CURDIR)/debian/libthrift-perl/usr/local/lib
-	rmdir $(CURDIR)/debian/libthrift-perl/usr/local
-
 	dh_install -i

 install-arch:
@@ -154,7 +125,6 @@ install-arch:
 	# Python
 	cd $(CURDIR)/lib/py && \
 	python2 setup.py install --install-layout=deb --no-compile --root=$(CURDIR)/debian/python-thrift && \
-	python2-dbg setup.py install --install-layout=deb --no-compile --root=$(CURDIR)/debian/python-thrift-dbg && \
 	python3 setup.py install --install-layout=deb --no-compile --root=$(CURDIR)/debian/python3-thrift && \
 	python3-dbg setup.py install --install-layout=deb --no-compile --root=$(CURDIR)/debian/python3-thrift-dbg

@@ -174,11 +144,6 @@ install-arch:
 	find $(CURDIR)/debian/python3-thrift-dbg -name "*.egg-info" -print0 | xargs -0 rm -rf
 	find $(CURDIR)/debian/python3-thrift-dbg -depth -type d -empty -exec rmdir {} \;

-	# PHP
-	mkdir -p $(CURDIR)/debian/php5-thrift
-	cd $(CURDIR)/lib/php && \
-		$(MAKE) DESTDIR=$(CURDIR)/debian/php5-thrift install
-
 	# C++ and C (glib)
 	mkdir -p $(CURDIR)/debian/tmp; \
 	cd $(CURDIR)/lib/cpp && \
--
2.17.1

