diff --git a/debian/rules b/debian/rules
index 3a50319ee..4b2a05932 100755
--- a/debian/rules
+++ b/debian/rules
@@ -65,6 +65,7 @@ build-compiler-stamp: configure-stamp
 
 build-libcpp-stamp: configure-stamp build-compiler-stamp
 	# Compile C++ library
+	if [ -f /usr/lib/arm-linux-gnueabihf/libboost_unit_test_framework.a ]; then sudo ln -s /usr/lib/arm-linux-gnueabihf/libboost_unit_test_framework.a /usr/lib/libboost_unit_test_framework.a ; fi
 	$(MAKE) -C $(CURDIR)/lib/cpp
 	touch $@
 
