diff --git a/CMakeLists.txt b/CMakeLists.txt
index fa562dd3..8635ba15 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -33,6 +33,7 @@ set(LIBYANG_MICRO_SOVERSION 2)
 set(LIBYANG_SOVERSION_FULL ${LIBYANG_MAJOR_SOVERSION}.${LIBYANG_MINOR_SOVERSION}.${LIBYANG_MICRO_SOVERSION})
 set(LIBYANG_SOVERSION ${LIBYANG_MAJOR_SOVERSION})
 
+set(CMAKE_INSTALL_PREFIX /usr)
 # set default build type if not specified by user
 if(NOT CMAKE_BUILD_TYPE)
     set(CMAKE_BUILD_TYPE debug)
diff --git a/packages/debian.control.in b/packages/debian.control.in
index da6588b9..fb3ede48 100644
--- a/packages/debian.control.in
+++ b/packages/debian.control.in
@@ -53,3 +53,15 @@ Depends: python3-yang@PACKAGE_PART_NAME@ (=@LIBYANG_VERSION@)
 Section: debug
 Architecture: any
 Description: Debug symbols of python3 bidings of libyang library.
+
+Package: python2-yang@PACKAGE_PART_NAME@
+Depends: @PACKAGE_NAME@ (=@LIBYANG_VERSION@), libyang-cpp@PACKAGE_PART_NAME@ (=@LIBYANG_VERSION@)
+Section: libs
+Architecture: any
+Description: Bindings of libyang library to python2 language.
+
+Package: python2-yang@PACKAGE_PART_NAME@-dbg
+Depends: python2-yang@PACKAGE_PART_NAME@ (=@LIBYANG_VERSION@)
+Section: debug
+Architecture: any
+Description: Debug symbols of python2 bidings of libyang library.
diff --git a/packages/debian.python2-yang.install b/packages/debian.python2-yang.install
new file mode 100644
index 00000000..14ce2f3c
--- /dev/null
+++ b/packages/debian.python2-yang.install
@@ -0,0 +1 @@
+usr/lib/python2.7/dist-packages/*
diff --git a/packages/debian.rules.in b/packages/debian.rules.in
index d565819e..e92fe4a1 100644
--- a/packages/debian.rules.in
+++ b/packages/debian.rules.in
@@ -9,10 +9,14 @@ export DH_VERBOSE=1
 override_dh_strip:
 	dh_strip -plibyang@PACKAGE_PART_NAME@  --dbg-package=libyang@PACKAGE_PART_NAME@-dbg
 	dh_strip -plibyang-cpp@PACKAGE_PART_NAME@ --dbg-package=libyang-cpp@PACKAGE_PART_NAME@-dbg
+	dh_strip -ppython2-yang@PACKAGE_PART_NAME@ --dbg-package=python2-yang@PACKAGE_PART_NAME@-dbg
 	dh_strip -ppython3-yang@PACKAGE_PART_NAME@ --dbg-package=python3-yang@PACKAGE_PART_NAME@-dbg
 
 override_dh_auto_configure:
-	cmake -DCMAKE_INSTALL_PREFIX:PATH=/usr -DCMAKE_BUILD_TYPE:String="@BUILD_TYPE@" -DGEN_LANGUAGE_BINDINGS=ON .
+	cmake -DCMAKE_INSTALL_PREFIX:PATH=/usr -DCMAKE_BUILD_TYPE:String="@BUILD_TYPE@" -DENABLE_LYD_PRIV=ON -DGEN_LANGUAGE_BINDINGS=ON -DGEN_PYTHON_VERSION=2 .
+
+override_dh_makeshlibs:
+	dh_makeshlibs -Xextensions -Xuser_types
 
 override_dh_auto_test:
 	ctest --output-on-failure
diff --git a/packages/libyang.dsc.in b/packages/libyang.dsc.in
index fdfa402b..f75ba184 100644
--- a/packages/libyang.dsc.in
+++ b/packages/libyang.dsc.in
@@ -1,10 +1,10 @@
 Format: 3.0 (quilt)
 Source: @PACKAGE_NAME@
-Binary: @PACKAGE_NAME@, @PACKAGE_NAME@-dbg, @PACKAGE_NAME@-dev, libyang-cpp@PACKAGE_PART_NAME@, libyang-cpp@PACKAGE_PART_NAME@-dev, libyang-cpp@PACKAGE_PART_NAME@-dbg, python3-yang@PACKAGE_PART_NAME@, python3-yang@PACKAGE_PART_NAME@-dbg
+Binary: @PACKAGE_NAME@, @PACKAGE_NAME@-dbg, @PACKAGE_NAME@-dev, libyang-cpp@PACKAGE_PART_NAME@, libyang-cpp@PACKAGE_PART_NAME@-dev, libyang-cpp@PACKAGE_PART_NAME@-dbg, python3-yang@PACKAGE_PART_NAME@, python3-yang@PACKAGE_PART_NAME@-dbg python2-yang@PACKAGE_PART_NAME@, python2-yang@PACKAGE_PART_NAME@-dbg
 Maintainer: CESNET <rkrejci@cesnet.cz>
 Version: @LIBYANG_VERSION@
 Architecture: any
 Standards-Version: 3.8.2
 Homepage: https://github.com/CESNET/libyang
 Vcs-Git: https://github.com/CESNET/libyang
-Build-Depends: debhelper (>= 9), make, gcc, doxygen, cmake, pkg-config, libpcre3-dev, libcmocka-dev, python3-dev, g++, swig (>= 3.0.12)
+Build-Depends: debhelper (>= 9), make, gcc, doxygen, cmake, pkg-config, libpcre3-dev, libcmocka-dev, python3-dev, python2-dev, g++, swig (>= 3.0.12)
diff --git a/packages/libyang.spec.in b/packages/libyang.spec.in
index 6a4ac615..6939f028 100644
--- a/packages/libyang.spec.in
+++ b/packages/libyang.spec.in
@@ -46,6 +46,8 @@ BuildRequires:  python3-devel
 %else
 BuildRequires:  python34-devel
 %endif
+
+BuildRequires:  python2-devel
 %endif
 
 Conflicts: @CONFLICT_PACKAGE_NAME@ = @LIBYANG_MAJOR_VERSION@.@LIBYANG_MINOR_VERSION@
@@ -70,6 +72,11 @@ Summary:    Binding to python
 Requires:   libyang-cpp@PACKAGE_PART_NAME@ = %{version}-%{release}
 Requires:   %{name} = %{version}-%{release}
 
+%package -n python2-yang@PACKAGE_PART_NAME@
+Summary:    Binding to python
+Requires:   libyang-cpp@PACKAGE_PART_NAME@ = %{version}-%{release}
+Requires:   %{name} = %{version}-%{release}
+
 %description -n libyang-cpp@PACKAGE_PART_NAME@
 Bindings of libyang library to C++ language.
 
@@ -80,6 +87,10 @@ Headers of bindings to c++ language.
 Bindings of libyang library to python language.
 %endif
 
+%description -n python2-yang@PACKAGE_PART_NAME@
+Bindings of libyang library to python language.
+%endif
+
 %description devel
 Headers of libyang library.
 
@@ -167,4 +178,9 @@ make DESTDIR=%{buildroot} install
 %{_libdir}/python*
 %endif
 
+%files -n python2-yang@PACKAGE_PART_NAME@
+%defattr(-,root,root)
+%{_libdir}/python*
+%endif
+
 %changelog
diff --git a/packages/local-deb.sh.in b/packages/local-deb.sh.in
index 057bbc67..4318a49d 100755
--- a/packages/local-deb.sh.in
+++ b/packages/local-deb.sh.in
@@ -18,6 +18,7 @@ fi
 cp "@PROJECT_SOURCE_DIR@/packages/debian.libyang-dev.install" debian/@PACKAGE_NAME@-dev.install
 cp "@PROJECT_SOURCE_DIR@/packages/debian.libyang-cpp.install" debian/libyang-cpp@PACKAGE_PART_NAME@.install
 cp "@PROJECT_SOURCE_DIR@/packages/debian.libyang-cpp-dev.install" debian/libyang-cpp@PACKAGE_PART_NAME@-dev.install
+cp "@PROJECT_SOURCE_DIR@/packages/debian.python2-yang.install" debian/python2-yang@PACKAGE_PART_NAME@.install
 cp "@PROJECT_SOURCE_DIR@/packages/debian.python3-yang.install" debian/python3-yang@PACKAGE_PART_NAME@.install
 echo -e "@PACKAGE_NAME@ (@LIBYANG_VERSION@) stable; urgency=low\n" >debian/changelog
 git log -10 --pretty=format:'  * %s (%aN)%n' 2>/dev/null >>debian/changelog || echo -e "  * unknown changes \n" >>debian/changelog
