From 4bdd892662c08a396066ba6a1c55eac3f8aa0a5f Mon Sep 17 00:00:00 2001
From: Vadym Hlushko <vadymh@nvidia.com>
Date: Tue, 25 Jan 2022 12:59:40 +0000
Subject: [PATCH] [sflow] enabled drop monitor support for SONiC

diff --git a/src/Linux/hsflowd.c b/src/Linux/hsflowd.c
index a29da54..4b6acc1 100644
--- a/src/Linux/hsflowd.c
+++ b/src/Linux/hsflowd.c
@@ -1922,6 +1922,14 @@ extern "C" {
     sp->psample.ingress = YES;
     sp->psample.egress = YES;
     sp->psample.group = 1; // Ingress PSAMPLE group number. Expects egress on (group+1).
+    // drop-monitor support
+    myLog(LOG_INFO, "drop-monitor support for SONiC");
+    sp->dropmon.dropmon = YES;
+    sp->dropmon.start = NO;
+    sp->dropmon.limit = 1000;
+    sp->dropmon.sw = NO;
+    sp->dropmon.hw = YES;
+
 #endif /* HSP_LOAD_SONIC */
 
 #ifdef HSP_LOAD_XEN
diff --git a/src/Linux/scripts/hsflowd.conf.sonic b/src/Linux/scripts/hsflowd.conf.sonic
index e675730..0604c5d 100644
--- a/src/Linux/scripts/hsflowd.conf.sonic
+++ b/src/Linux/scripts/hsflowd.conf.sonic
@@ -4,6 +4,8 @@
 sflow {
   # sonic {} loaded automatically
   # psample {} loaded automatically
+  # packet drop monitoring:
+  dropmon { group=1 start=on limit=1000 sw=off hw=on }
   # ====== detect new interfaces ======
   refreshAdaptors=60
   # ======  Agent IP selection ======
