From 9f96e358fb5bdac8a6570dc3c79e9416e4f407ce Mon Sep 17 00:00:00 2001
From: Nobuaki Sukegawa <nsuke@apache.org>
Date: Mon, 22 Feb 2016 01:33:27 +0900
Subject: [PATCH] THRIFT-3577 assertion failed at line 512 of
 testcontainertest.c

---
 lib/c_glib/test/testcontainertest.c | 6 +++---
 1 file changed, 3 insertions(+), 3 deletions(-)

diff --git a/lib/c_glib/test/testcontainertest.c b/lib/c_glib/test/testcontainertest.c
index 852254b..1cbc55c 100644
--- a/lib/c_glib/test/testcontainertest.c
+++ b/lib/c_glib/test/testcontainertest.c
@@ -507,9 +507,9 @@ main(int argc, char *argv[])
 
     /* Make sure the server stopped only because it was interrupted (by the
        child process terminating) */
-    g_assert (g_error_matches (error,
-                               THRIFT_SERVER_SOCKET_ERROR,
-                               THRIFT_SERVER_SOCKET_ERROR_ACCEPT));
+    g_assert(!error || g_error_matches(error,
+                                       THRIFT_SERVER_SOCKET_ERROR,
+                                       THRIFT_SERVER_SOCKET_ERROR_ACCEPT));
 
     /* Free our resources */
     g_object_unref (server);
